File: CarePackage/Software/DnSpy.cs
namespace CarePackage.Software;

public class DnSpy : BaseSoftware
{
    public override string           Key            { get; set; } = "dnspy-ex";
    public override string           Name           { get; set; } = "dnSpy (Fork)";
    public override SoftwareCategory Category       { get; set; } = SoftwareCategory.Development;
    public override string           DownloadName   { get; set; } = "dnSpy-net-win64.zip";
    public override bool             IsArchive      { get; set; } = true;
    public override bool             ShouldCacheUrl { get; set; } = true;
    public override bool             RequiresAdmin  { get; set; } = false;
    public override Bitmap           Icon           { get; set; } = Resources.Icons.dnspy;
    public override string           Homepage       { get; set; } = "https://github.com/dnSpyEx/dnSpy";
    
    private readonly GitHubService _github;

    public DnSpy(IServiceProvider services)
    {
        _github = services.GetRequiredService<GitHubService>();
    }
    
    public override async Task<string> GetDownloadUrlAsync(CancellationToken ct)
    {
        var assets = await _github.GetLatestRepositoryReleaseAssetsAsync("dnSpyEx", "dnSpy", ct);
        var asset  = assets.FirstOrDefault(a => a.EndsWith("win64.zip"));
        
        DownloadUrlResolveException.ThrowIf(asset is null);

        return asset;
    }
}

File: packages/app/src/services/software/definitions/dnspy.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory } from 'shared';
import { GithubReleasesScraper } from '../githubReleasesScraper';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'dnspy-ex';
	public name = 'dnSpy (Fork)';
	public category = [SoftwareCategory.Development, SoftwareCategory.Utility];
	public downloadName = 'dnSpy-net-win64.zip';
	public isArchive = true;
	public shouldCacheUrl = true;
	public icon = 'dnspy.png';
	public homepage = 'https://github.com/dnSpyEx/dnSpy';

	public async resolveDownloadUrl() {
		const owner   = 'dnSpyEx';
		const repo    = 'dnSpy';
		const release = await GithubReleasesScraper.getRelease(owner, repo, r => r.endsWith('win64.zip'));
		if (release.isErr()) {
			return err(release.error);
		}

		return ok(release.value);
	}
}
